File: _examples/simple_job.go
package main

import (
	"context"
	"fmt"
	"log"
	"log/slog"
	"os"
	"os/signal"
	"syscall"
	"time"

	"github.com/icegatetech/jobmanager"
)

// This minimal example demonstrates how to configure and run a simple job with a single task.
// It sets up the necessary components: logger, storage (S3), job definition, and the job manager itself.

func main() {
	if err := runSimpleJob(); err != nil {
		log.Fatal(err)
	}
}

func runSimpleJob() error {
	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
	defer cancel()

	lgr := jobmanager.NewSlogLogger(
		slog.New(slog.NewTextHandler(os.Stdout, &slog.HandlerOptions{Level: slog.LevelDebug})),
	)

	// 1. Define the initial task
	taskDef, err := jobmanager.NewTaskDefinition("my task code", nil)
	if err != nil {
		return fmt.Errorf("failed to create task executor: %w", err)
	}
	// 2. Map executor for task
	taskExecutors := map[jobmanager.TaskCode]jobmanager.TaskExecutor{"my task code": taskExecutor}
	// 3. Define the job
	jobDef, err := jobmanager.NewJobDefinition(
		"my job code",
		[]jobmanager.TaskDefinition{taskDef},
		taskExecutors,
	)
	if err != nil {
		return fmt.Errorf("failed to create job: %w", err)
	}
	// 4. Get the jobs together
	jobDefs, err := jobmanager.NewJobDefinitions(jobDef)
	if err != nil {
		return fmt.Errorf("failed to create jobs: %w", err)
	}

	// 5. Initialize S3 Storage (MinIO)
	storage, err := jobmanager.NewS3Storage(
		ctx,
		jobmanager.S3StorageConfig{
			Endpoint:        "localhost:9000",
			AccessKeyID:     "minioadmin",
			SecretAccessKey: "minioadmin",
			BucketName:      "jobs",
			UseSSL:          false,
			BucketPrefix:    "jobs",
		},
		lgr,
		jobmanager.NewDisabledMetrics(),
		jobDefs,
		jobmanager.NewRetrier(jobmanager.RetrierConfig{}, lgr),
	)
	if err != nil {
		return fmt.Errorf("failed to create storage: %w", err)
	}

	// 6. Initialize JobsManager to manage all of them
	manager, err := jobmanager.NewJobsManager(storage, jobmanager.JobsManagerConfig{}, lgr, jobDefs)
	if err != nil {
		return fmt.Errorf("failed to create manager: %w", err)
	}

	// 7. Start (blocking)
	if err := manager.Start(ctx); err != nil {
		return fmt.Errorf("failed to start manager: %w", err)
	}

	return nil
}

func taskExecutor(ctx context.Context, task jobmanager.ImmutableTask, manager jobmanager.JobManager) error {
	// Simulate work
	time.Sleep(100 * time.Millisecond)

	return manager.CompleteTask(task.ID(), []byte("done"))
}

File: examples/simple_job.rs
// This minimal example demonstrates how to configure and run a simple job with a single task.
// It sets up the necessary components: logger, storage (S3), job definition, and the job manager
// itself.

#![allow(missing_docs)]

use std::{collections::HashMap, sync::Arc, time::Duration};

use chrono::Duration as ChronoDuration;
use jobmanager::{
    Error, JobCode, JobDefinition, JobRegistry, JobStateCodecKind, JobsManager, JobsManagerConfig, Metrics, S3Storage,
    S3StorageConfig, TaskCode, TaskDefinition, TaskExecutorFn,
};

#[allow(dead_code)]
#[tokio::main]
async fn main() -> Result<(), Error> {
    run_simple_job(JobStateCodecKind::Json, "simple-json".to_string()).await
}

pub async fn run_simple_job(codec: JobStateCodecKind, bucket_prefix: String) -> Result<(), Error> {
    // Initialize tracing/logging
    tracing_subscriber::fmt()
        .with_target(false)
        .with_env_filter("jobmanager=debug")
        .init();

    tracing::info!("Starting simple job example");

    // 1. Define the initial task
    let task_def = TaskDefinition::new(TaskCode::new("my task code"), Vec::new(), ChronoDuration::seconds(5))?;

    // 2. Map executor for task
    let task_executor: TaskExecutorFn = Arc::new(|task, manager, _cancel_token| {
        // Extract task ID before async block to avoid Send trait issues
        let task_id = *task.id();

        Box::pin(async move {
            tracing::info!("Executing task: {}", task_id);

            // Simulate work
            tokio::time::sleep(Duration::from_millis(100)).await;

            tracing::info!("Task completed: {}", task_id);
            manager.complete_task(&task_id, b"done".to_vec())
        })
    });

    let mut task_executors = HashMap::new();
    task_executors.insert(TaskCode::new("my task code"), task_executor);

    // 3. Define the job
    let job_def = JobDefinition::new(JobCode::new("simple job"), vec![task_def], task_executors)?;

    // 4. Create job definitions
    let job_registry = Arc::new(JobRegistry::new(vec![job_def])?);

    // 5. Initialize S3 Storage
    let storage = S3Storage::new(
        S3StorageConfig::new(
            "http://localhost:9000",
            "rustfsadmin",
            "rustfsadmin",
            "jobs",
            "us-east-1",
        )
        .with_bucket_prefix(bucket_prefix)
        .with_job_state_codec(codec),
        job_registry.clone(),
        Metrics::new_disabled(),
    )
    .await?;

    // 6. Initialize JobsManager
    let manager = JobsManager::new(
        Arc::new(storage),
        JobsManagerConfig::default(),
        Arc::clone(&job_registry),
        Metrics::new_disabled(),
    )?;

    // 7. Start the manager with signal handling
    tracing::info!("Starting job manager (press Ctrl+C to stop)");

    let handle = manager.start()?;
    tokio::signal::ctrl_c()
        .await
        .map_err(|e| Error::Other(format!("ctrl_c error: {e}")))?;
    tracing::info!("Received interrupt signal, shutting down...");
    handle.shutdown().await?;

    tracing::info!("Job manager stopped");
    Ok(())
}
